src/format_string.cpp

```diff
@@ -39,6 +41,37 @@ ArgumentType get_expected_argument_type(const std::string &fmt)
   return ArgumentType::UNKNOWN;
 }
 
+// Returns a vector of (token, type) tuples given a format string
+std::vector<std::tuple<std::string, Type>> get_token_types(
```

Comment:
nit: Maybe a few unit tests for this would be good.

Reply:
Took a look. Could probably be done, but would need to pull this out of anon namespace. We also already have quite a bit of coverage in semantic analyser tests. See `semantic_analyser.printf_*` set of tests. So I think coverage is likely sufficient here. WDYT?